src/security/Session.cc

```diff
@@ -136,6 +136,10 @@ CreateSession(const Security::ContextPointer &ctx, const Comm::ConnectionPointer
         const int fd = conn->fd;
 
 #if USE_OPENSSL
+		#if USE_OPENSSL_KTLS
+	    SSL_set_options(session.get(), SSL_OP_ENABLE_KTLS);
```

Comment:
Security::PeerOptions::updateSessionOptions() also sets options. If SSL_OP_ENABLE_KTLS option can be set by adding a new ssl_options entry (and, hence, recognizing a new peer-specific setting in squid.conf), please do so instead of hard-coding this setting (for all peers)?

If the above or a similar plan works, we may be even be able to avoid introducing USE_OPENSSL_KTLS!

Reply:
Just reverting this may not be a good idea.

When using KTLS, perhaps the upper layer OpenSSL function (SSL_*) does not want this Ssl::ServerBio::read function to read more data from the socket (file descriptor) than the OpenSSL function is expecting because at some specific point the OpenSSL function changes the socket option.

https://github.com/openssl/openssl/blob/5963aa8c196d7c5a940a979299a07418527932af/include/internal/ktls.h#L70C3-L70C3


I cannot reproduce the issue that I was encountering while prototyping though, depending on the timing of packet arrival, the current (old/official) code potentially read more than expected data from the socket, which has been OK for non-ktls session but not for ktls.

Perhaps we have to consider some alternative way.

Reply:
supported options=ENABLE_KTLS in squid.conf

https://github.com/squid-cache/squid/pull/1629/commits/14c81ec77ce855444453a6b2d822da7f549d4acd